relayer/workers/worker.go

```diff
@@ -99,23 +108,41 @@ func (wp WorkerPool) run(ctx context.Context, log *logrus.Entry) error {
 		factory := f
 
 		eg.Go(func() error {
+			restarts := 0
+
 			for {
-				worker, err := factory()
+				worker, config, err := factory()
 				if err != nil {
 					// It is unrecoverable if we cannot construct one of our workers
 					return err
 				}
 
-				log.WithField("worker", worker.Name()).Debug("Starting worker")
-				err = wp.runWorker(ctx, worker, log)
-				log.WithField("worker", worker.Name()).Debug("Worker terminated")
+				if !config.Enabled {
+					return nil
+				}
+
+				log.WithFields(logrus.Fields{
+					"restarts": restarts,
+					"worker":   worker.Name(),
+				}).Debug("Starting worker")
+				err = wp.runWorker(ctx, worker)
+
+				if err == WorkerDeadlocked {
+					log.WithField(
+						"worker",
+						worker.Name(),
+					).Error("The worker's goroutines are deadlocked. Please fix")
+					return onDeadlock()
+				} else {
+					log.WithError(err).WithField("worker", worker.Name()).Debug("Worker terminated")
+				}
 
 				select {
 				case <-ctx.Done():
 					return ctx.Err()
 				default:
-					// TODO: instead retry with backoff up to X retries
-					return err
+					<-time.After(time.Duration(config.RestartDelay) * time.Second)
```

Comment:
I figured this is good enough for now. We can do backoffs and limit restarts if we see we need it